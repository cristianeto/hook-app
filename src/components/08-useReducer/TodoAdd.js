import React from "react";
import { useForm } from "../../hooks/useForm";

const TodoAdd = ({ onAddTodo }) => {
  //? usnado el hook personalizado llamado useForm
  //const [formValues, handleInputCange] = useForm({
  const [{ description }, handleInputCange, reset] = useForm({
    description: "",
  });

  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim().length <= 1) return;

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };

    onAddTodo(newTodo);
    reset();
  };
  return (
    <>
      <h4>Agregar TODO</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          autoComplete="off"
          className="form-control"
          name="description"
          onChange={handleInputCange}
          placeholder="description"
          type="text"
          value={description}
        />
        <button
          type="submit"
          className="btn btn-outline-primary mt-1 btn-block"
        >
          Agregar
        </button>
      </form>
    </>
  );
};

export default TodoAdd;
