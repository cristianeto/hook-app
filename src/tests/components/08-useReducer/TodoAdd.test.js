import { shallow } from "enzyme";
import TodoAdd from "../../../components/08-useReducer/TodoAdd";

describe("Testing for <TodoAdd/>", () => {
  const handleAddTodo = jest.fn();
  const wrapper = shallow(<TodoAdd onAddTodo={handleAddTodo} />);

  test("should return <TodoAdd/> component correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should not allow submit when is empty input", () => {
    const formSubmit = wrapper.find("form").prop("onSubmit");
    formSubmit({ preventDefault() {} });

    expect(handleAddTodo).toHaveBeenCalledTimes(0);
  });

  test("should call onAddTodo function once", () => {
    /**Arrange */
    const value = "Aprender React";
    wrapper.find("input").simulate("change", {
      target: {
        value,
        name: "description",
      },
    });
    const formSubmit = wrapper.find("form").prop("onSubmit");
    /** ? Act*/
    formSubmit({ preventDefault() {} });
    /* Assert ? llamada a la function onAddTodo 1 vez*/
    expect(handleAddTodo).toHaveBeenCalledTimes(1);
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      desc: value,
      done: false,
    });
    expect(wrapper.find("input").prop("value")).toBe(""); //?se llamo a la funcion reset, se comprueba q se vació o limpió el input
  });
});
