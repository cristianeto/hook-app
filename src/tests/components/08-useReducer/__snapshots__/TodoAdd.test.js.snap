// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing for <TodoAdd/> should return <TodoAdd/> component correctly 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="description"
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
